rocAL/source/meta_data/webdataset_meta_data_reader.cpp

```diff
@@ -0,0 +1,414 @@
+        _wds_shards.reserve(entry_name_list.size());
+        for (auto &path : entry_name_list)
+            _wds_shards.emplace_back(std::make_unique<std::ifstream>( folder_path + path, std::ios::binary));
+    }
+    closedir(_sub_dir);
+
+    std::vector<SampleDescription> unfiltered_samples;
+    std::vector<ComponentDescription> unfiltered_components;
+
+    for (unsigned wds_shard_index = 0; wds_shard_index < entry_name_list.size();
+         ++wds_shard_index) {
+        unfiltered_samples.resize(0);
+        unfiltered_components.resize(0);
+        if (_index_paths.size() == 0)
+            parse_tar_files(unfiltered_samples, unfiltered_components,
+                            _wds_shards[wds_shard_index]);
+        else
+            parse_index_files(unfiltered_samples, unfiltered_components,
+                              _folder_path + _index_name_list[wds_shard_index]);
+
+        // After parsing add the contents to the map
+        for (auto &sample : unfiltered_samples) {
+            if (_missing_component_behaviour == MissingComponentsBehaviour::MISSING_COMPONENT_EMPTY) {  // empty_outputs
+                AsciiValues ascii_values;
+                ascii_values.resize(_ext_map.size());
+                std::string last_file_name;
+                for (auto &component : sample.components) {
+                    if (!isJPEG(component.ext)) {  // Add more components as we encounter
+                        _wds_shards[wds_shard_index]->seekg(component.offset, std::ios::beg);
+                        AsciiComponent ascii_component = std::make_shared<std::vector<uint8_t>>(component.size);
+                        _wds_shards[wds_shard_index]->read(reinterpret_cast<char*>(ascii_component->data()), component.size);
+                        ascii_values.at(_ext_map[component.ext]) = ascii_component;
+                    }
+                    last_file_name = component.filename;
+                }
+                add(last_file_name, ascii_values);
+                ascii_values.clear();
+            } else {
+                auto skip_sample = false;
+                AsciiValues ascii_values;
+                ascii_values.resize(_ext_map.size());
+                std::string last_file_name;
+                for (auto &component : sample.components) {
+                    if (!isJPEG(component.ext)) {  // Add more components as we encounter
+                        _wds_shards[wds_shard_index]->seekg(component.offset, std::ios::beg);
+                        AsciiComponent ascii_component = std::make_shared<std::vector<uint8_t>>(component.size);
+                        _wds_shards[wds_shard_index]->read(reinterpret_cast<char*>(ascii_component->data()), component.size);
+                        ascii_values.at(_ext_map[component.ext]) = ascii_component;
+                    }
+                    last_file_name = component.filename;
+                }
+                for (auto &ascii_component : ascii_values) {
```

Comment:
can't you do this in the previous loop

Reply:
No Rajy, The vector is initialized with nullptr's at the beginning. And then we add all the components present in the sample. 
When any component goes MISSING, we look for if the component is nullptr, and handle the behavior accordingly.